### ISSUE TITLE
Nullref in HttpDiagnosticListenerImplBase.ProcessExceptionEvent

### ISSUE THREAD

**Issue by gregkalapos** [elastic/apm-agent-dotnet#443]
From https://discuss.elastic.co/t/apm-bug-causes-null-pointer-for-web-requests-in-net-4-6-1/195923

**Comment by SergeyKleyman** [comment]
It might be also worth improving logging while we at it. At the moment it would have been not very useful for this bug:
```
[2019-08-20 11:56:43][Trace] - {HttpDiagnosticListenerImplBase} Processing exception event... Request URL: https://xxxx/api/georef/v1/regions/details?orqId=Cs_6a91a9c4-e40c-4457-89f2-06570b3f9805
Application startup exception: System.Net.WebException: The request was aborted: The request was canceled. ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at Elastic.Apm.DiagnosticListeners.HttpDiagnosticListenerImplBase`2.ProcessExceptionEvent(Object eventValue, Uri requestUrl)
   at Elastic.Apm.DiagnosticListeners.HttpDiagnosticListenerImplBase`2.OnNext(KeyValuePair`2 kv)
   at System.Diagnostics.DiagnosticListener.Write(String name, Object value)
   at System.Diagnostics.HttpHandlerDiagnosticListener.RaiseResponseEvent(HttpWebRequest request, HttpStatusCode statusCode, WebHeaderCollection headers)
   at System.Diagnostics.HttpHandlerDiagnosticListener.HttpWebRequestArrayList.RemoveAt(Int32 index)
   at System.Net.Connection.ReadStartNextRequest(WebRequest currentRequest, ConnectionReturnResult& returnResult)
   at System.Net.ConnectStream.CallDone(ConnectionReturnResult returnResult)
   at System.Net.HttpWebResponse..ctor(Uri responseUri, KnownHttpVerb verb, CoreResponseData coreData, String mediaType, Boolean usesProxySemantics, DecompressionMethods decompressionMethod, Boolean isWebSocketResponse, String connectionGroupName)
   at System.Net.HttpWebRequest.SetResponse(CoreResponseData coreResponseData)
   --- End of inner exception stack trace ---
   at System.Net.HttpWebRequest.GetResponse()
   at ServiceStack.ServiceClientBase.Send[TResponse](String httpMethod, String relativeOrAbsoluteUrl, Object request)
```
We should log `eventValue` paramter to `ProcessExceptionEvent` and maybe for `Start` and `Stop` events as well (using `ToString` should be enough since `ToString` is overridden by compiler for anonymous types).

**Comment by SergeyKleyman** [comment]
Regarding the bug itself - the root cause is that I misinterpreted what `"System.Net.Http.Desktop.HttpRequestOut.Ex.Stop"` event means. I thought `Ex` is short for Exception but according to 
https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/HttpHandlerDiagnosticListener.cs#L560
`"System.Net.Http.Desktop.HttpRequestOut.Ex.Stop"` event is the same as `"System.Net.Http.Desktop.HttpRequestOut.Stop"` event but with different payload.

I don't see any equivalent of [.NET Core's `"System.Net.Http.Exception"` event](https://github.com/dotnet/corefx/blob/master/src/System.Net.Http/src/System/Net/Http/DiagnosticsHandlerLoggingStrings.cs#L16) for .NET Framework, at least not in [src/System/Diagnostics/HttpHandlerDiagnosticListener.cs](https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/HttpHandlerDiagnosticListener.cs#L797). [Here](https://github.com/dotnet/corefx/blob/master/src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/HttpHandlerDiagnosticListener.cs#L548) comment mentions exception as possible outcome of HTTP call but I don't see any attempt to extract that exception and pass it to the DiagnosticListeners. Further investigation is required.

This bug is related to https://github.com/elastic/apm-agent-dotnet/issues/291 - now it's clear why our DiagnosticListener did not get events when `HttpClient` threw exception - `"System.Net.Http.Desktop.HttpRequestOut.Ex.Stop"` event is not for exceptions.

**Comment by gregkalapos** [comment]
> It might be also worth improving logging while we at it. At the moment it would have been not very useful for this bug:
> 
> ```
> [2019-08-20 11:56:43][Trace] - {HttpDiagnosticListenerImplBase} Processing exception event... Request URL: https://xxxx/api/georef/v1/regions/details?orqId=Cs_6a91a9c4-e40c-4457-89f2-06570b3f9805
> Application startup exception: System.Net.WebException: The request was aborted: The request was canceled. ---> System.NullReferenceException: Object reference not set to an instance of an object.
>    at Elastic.Apm.DiagnosticListeners.HttpDiagnosticListenerImplBase`2.ProcessExceptionEvent(Object eventValue, Uri requestUrl)
>    at Elastic.Apm.DiagnosticListeners.HttpDiagnosticListenerImplBase`2.OnNext(KeyValuePair`2 kv)
>    at System.Diagnostics.DiagnosticListener.Write(String name, Object value)
>    at System.Diagnostics.HttpHandlerDiagnosticListener.RaiseResponseEvent(HttpWebRequest request, HttpStatusCode statusCode, WebHeaderCollection headers)
>    at System.Diagnostics.HttpHandlerDiagnosticListener.HttpWebRequestArrayList.RemoveAt(Int32 index)
>    at System.Net.Connection.ReadStartNextRequest(WebRequest currentRequest, ConnectionReturnResult& returnResult)
>    at System.Net.ConnectStream.CallDone(ConnectionReturnResult returnResult)
>    at System.Net.HttpWebResponse..ctor(Uri responseUri, KnownHttpVerb verb, CoreResponseData coreData, String mediaType, Boolean usesProxySemantics, DecompressionMethods decompressionMethod, Boolean isWebSocketResponse, String connectionGroupName)
>    at System.Net.HttpWebRequest.SetResponse(CoreResponseData coreResponseData)
>    --- End of inner exception stack trace ---
>    at System.Net.HttpWebRequest.GetResponse()
>    at ServiceStack.ServiceClientBase.Send[TResponse](String httpMethod, String relativeOrAbsoluteUrl, Object request)
> ```
> 
> We should log `eventValue` paramter to `ProcessExceptionEvent` and maybe for `Start` and `Stop` events as well (using `ToString` should be enough since `ToString` is overridden by compiler for anonymous types).

Don't we log those already [here](https://github.com/elastic/apm-agent-dotnet/blob/master/src/Elastic.Apm/DiagnosticListeners/HttpDiagnosticListenerImplBase.cs#L61)? - that's the caller method. 

Also, logs were attached on the linked discuss topic, and those logs are:
```
[2019-08-20 11:56:43][Trace] - {HttpDiagnosticListenerImplBase} Called with key: `
System.Net.Http.Desktop.HttpRequestOut.Ex.Stop', value: `{ Request = System.Net.HttpWebRequest, StatusCode = Unauthorized, Headers = Connection: keep-alive
...
```

I think that was exactly what we needed - it shows that the `value` has 3 properties `Request`, `StatusCode` and `Headers` and does not have the `Exception` property we try to use. Or do I miss something? I think the logs were fine.

**Comment by SergeyKleyman** [comment]
You're right - we already log it.

**Comment by gregkalapos** [comment]
Fixed in #450
